noir-projects/aztec-nr/aztec/src/state_vars/array_dynamic_public.nr

```diff
@@ -0,0 +1,91 @@
+use crate::{
+    context::Context, oracle::{storage::{storage_read, storage_write}}, state_vars::{storage::Storage,map::Map}
+};
+use dep::protocol_types::{hash::pedersen_hash, constants::INITIALIZATION_SLOT_SEPARATOR, traits::{Deserialize, Serialize}};
+
+struct PublicDynamicArray<T> {
+  context: Context,
+  map: Map<u64, T>,
+  storage_slot: Field,
+}
+
+impl<T> Storage<T> for PublicDynamicArray<T> {}
+
+impl<T> PublicDynamicArray<T> {
+  pub fn new(
+    context: Context,
+    storage_slot: Field,
+    state_var_constructor: fn(Context, Field) -> T
+  ) -> Self {
+    assert(storage_slot != 0, "Storage slot 0 not allowed. Storage slots must start from 1.");
+
+    PublicDynamicArray { 
+      context, 
+      storage_slot, 
+      map: Map::new(
+        context,
+        storage_slot + INITIALIZATION_SLOT_SEPARATOR,
```

Comment:
Storage slots cannot be derived using fixed offsets in public or we won't be able to assign them automatically (we don't know the size of the stuff we're storing here). 

We could use the storage slot of the map to store the length, since the map doesn't occupy it by itself. However, this length has to be initialized to 1, or we will generate an invalid key when the map derives the slot for the first item